### DISCUSSION TITLE
2024 Yearly Planning(Want  Your Feedback)

### DISCUSSION THREAD

**Discussion by git-hulk** [apache/kvrocks#1974] [General]
Like [2023 Planning](https://github.com/apache/kvrocks/discussions/1226), the community would like to hear from you before starting the yearly planning. Don't hesitate to leave any comments or features you want.

And we would bring them into [Kvrocks 2024 Roadmap](https://github.com/notifications?query=is%3Aunread) and wait for contributors to take it.

Thank you!

**Comment by iushas** [comment]
实现类似 mysql两阶段提交的支持，增加 binlog，可以通过 binlog 实现，
1、kvrocks 全量 + 增量 准确位点或时间的备份恢复。  
2、快速的主从同步，避免psync信息没有时，全量拷贝数据

**Comment by mapleFU** [comment]
I hope we can have a random testing tool in 2024. We could:
1. Testing the bug in commands
2. Testing the bug in synchronization

Maybe it could also testing the compatibility with redis. I'm not familiar with Redis, so don't know if there're any existing tools

**Reply by git-hulk** [reply]
> Maybe it could also testing the compatibility with redis. I'm not familiar with Redis, so don't know if there're any existing tools

https://github.com/tair-opensource/compatibility-test-suite-for-redis.git can resolve this issue.

**Reply by mapleFU** [reply]
Oops. I think my main purpose is that it can detect our own bugs. Do you know any fuzzing or random testing redis tools?

**Reply by git-hulk** [reply]
Not sure if @enjoy-binbin has any input about this.

**Reply by enjoy-binbin** [reply]
redis itself had some random fuzzing on its TCL test cases, to verify its internal implementation (such as some encoding related cases.)

however, i don’t know much about the external aspects. As far as I know, some of them are just randomly pouring in a large number of commands (and randomly generated correct or incorrect args) to see if the server will crash.

**Reply by caipengbo** [reply]
> Testing the bug in synchronization

Maybe we can introduce [sync point](https://github.com/facebook/rocksdb/blob/main/test_util/sync_point.h) ? @mapleFU

**Reply by jihuayu** [reply]
Maybe we can use kvrocks test units to test Redis.
It can prove the correctness of our test cases.

**Reply by mapleFU** [reply]
> Maybe we can introduce [sync point](https://github.com/facebook/rocksdb/blob/main/test_util/sync_point.h) ?

I've port a part of SyncPoint but I think it's used to test concurrency: https://github.com/mapleFU/sync_point

> Maybe we can use kvrocks test units to test Redis. It can prove the correctness of our test cases.

Maybe. I think our integration test are trivial and hard to failed on that

**Comment by i18nsite** [comment]
1. 支持 resp3，并且支持lua脚本 setResp
2. 完整支持redis的stream ，现在实现不完整，很多用法用不起来
3. 集群部署希望支持用mysql替代etcd
4. lua脚本在mac docker容器无法实时输出的bug不知道能不能复现修一下
5. 能不能支持类似oeacnbase的租户和配额，有了这个就可以做成云服务卖
___
1. Support resp3 and lua script setResp
2. Full support for redis stream, now the implementation is incomplete, many uses can not be used
3. Cluster deployment supports mysql instead of etcd
4. lua script in mac docker container can not be output in real time bug do not know can be repeated to fix it
5. Can we support tenants and quotas similar to oeacnbase, with which we can sell cloud services

**Reply by jihuayu** [reply]
Stream has many commands that need to be impl. No one has claimed the relevant task for a long time.
Perhaps the relevant issues do not have the "good first issues" label. 
Or is the data structure of the stream too complex for developers to understand?
- https://github.com/apache/kvrocks/issues/1585

**Comment by xiaobiaozhao** [comment]
New function timeseries

https://redis.io/docs/data-types/timeseries/

https://github.com/apache/kvrocks/discussions/1876

**Comment by ediphy-dwild** [comment]
in terms of documentation / tutorials - it would be really useful to have an article on how to set up a kvrocks cluster from scratch. the existing articles that i can find assume familiarity with various redis cluster setups - which is not always the case. when looking at kvrocks as a candidate for use, this was one of the areas of concern for us.

in particular, its not clear how to set up a cluster without operator intervention at startup.
eg, should people use the web ui (sorry i cant remember the name of it now, and i cant easily find a reference to the name of it), or does the cluster setup require redis-cli commands to be entered interactively (this is an issue for a lot of automated devops setups, where people expect to be able to set up and tear down a cluster using just docker-compose / kubernetes yaml files) etc

if there was a docker-compose file available to allow setup and tear down, it would be very useful, even for people that do not use that particular orchestration tool, as it would be forced to include everything required to set up a cluster in a repeatable manner without operator intervention.

**Reply by xiaobiaozhao** [reply]
I thought we could write a documentation on how to use the kvrocks_controller component for a productionable cluster.

**Reply by git-hulk** [reply]
Yes, it's a bit hard to use https://github.com/apache/kvrocks-controller without UI, and now we're building this. Hope it can bring more smoother experience for users in the next coming release.

**Comment by jcy1001** [comment]
the most import one I think  is kvrocks-controller  ,use it  incremental backup ,and  restore and  repair data corrupt ,and so, contoller use inner datastore not etcd ,mysql zookeeper like .

**Reply by git-hulk** [reply]
One of the Kvrocks's GSoC project ideas is working on this.

**Comment by jcy1001** [comment]
kvrocks-controller提供一个web UI界面，上面可以配置集群故障出现时候进行告警和自动故障转移功能，比如遇到rocksdb的sst失败写入失败，可以在这个UI界面进行配置后，收到告警通知，和进行主从切换这些操作

**Reply by jcy1001** [reply]
kvrocks能否增加强一致性的，我们有的业务场景非常需要强一致性，这样可以移到kvrocks上来做这块

**Comment by jcy1001** [comment]


**Reply by git-hulk** [reply]
@jcy1001 Thanks for your proposal, maybe we can cast this into an issue to see if any one is interested in doing this.

**Comment by kinoute** [comment]
* A Helm Chart to deploy easily KVRocks (cluster?) on Kubernetes with master and replicas such as ElasticSearch
* A way to remove keys when `max-db-size` is reached (like LRU)

**Reply by git-hulk** [reply]
@kinoute Thanks for your feedback. I will add them into the yearly plan.

**Comment by sergerus** [comment]
Please improve release process. Give advance notice of the contents (bugfixes, features) of the upcoming release. If possible freeze content of upcoming release. Make milestones RC1, RC2 etc.

**Reply by PragmaTwice** [reply]
Hi @sergerus , thank you for your suggestion.

For releasing, we are following the official ASF release process. You can check https://kvrocks.apache.org/community/create-a-release for more information.

**Comment by git-hulk** [comment]
Add the support of ACL user, refer: https://github.com/apache/kvrocks/discussions/2274
